examples/hhl.py

```diff
@@ -194,73 +201,221 @@ def _ancilla_rotation(self, k):
+        """Simulates the HHL circuit with amplitude amplification.
+
+        Amplitude amplification is applied to the HHL circuit until the ancilla is measured as 1.
+        Afterwards X, Y and Z are measured on the memory qubit.
+
+        Args:
+            total_estimates: The number of estimates to obtain for each X, Y and Z measurement.
+        """
+        simulator = cirq.Simulator()
+
+        # Cases for measuring X, Y, and Z (respectively) on the memory qubit.
+        params = [
+            {'exponent': 0.5, 'phase_exponent': -0.5},
+            {'exponent': 0.5, 'phase_exponent': 0},
+            {'exponent': 0, 'phase_exponent': 0},
+        ]
 
-    for label, result in zip(('X', 'Y', 'Z'), list(results)):
-        # Only select cases where the ancilla is 1.
-        # TODO: optimize using amplitude amplification algorithm.
-        # Github issue: https://github.com/quantumlib/Cirq/issues/2216
-        expectation = 1 - 2 * np.mean(result.measurements['m'][result.measurements['a'] == 1])
-        print(f'{label} = {expectation}')
+        results = []
+        runs = []
+        for param in params:
+            resolver = cirq.ParamResolver(param)
+            results_for_param = []
+            repetitions = total_estimates
+            num_runs = 0  # number of calls to simulator.run
+            # We use the algorithm "Quantum Searching Without Knowing Success Probability II"
+            # from Kaye et al., Section 8.4.
+            # We perform 9 iterations, which suffices if the success probability p of a single run
+            # without amplification is at least 1/2^10.
+            for l in range(1, 10):
+                M = 2**l
+                # Apply amplitude amplification a random number of times between 0 and M-1.
+                # For M large enough, doing this twice increases the success probability to
+                # 3/4 - O(1/(M*θ)), where θ is the angle such that sin^2(θ) = p, the probability
+                # of success of a single run without amplification.
+                for _ in range(2):
+                    j = random.randint(0, M - 1)
+                    result = simulator.run(
+                        self.measure_circuit(self.amplitude_amplification(num_iterations=j)),
+                        resolver,
+                        repetitions,
+                    )
+                    num_runs += repetitions
+                    results_for_param.append(
+                        result.measurements['m'][result.measurements['a'] == 1]
+                    )
+                    repetitions -= (result.measurements['a'] == 1).sum()
+                    if repetitions <= 0:
+                        break
+                if repetitions <= 0:
+                    break
+
+            results.append(np.concatenate(results_for_param, axis=0))
+            runs.append(num_runs)
+
+        for label, result, num_runs in zip(('X', 'Y', 'Z'), results, runs):
```

Comment:
I think it might be better and more testable if we return a result object and then print it out after it returns.

Reply:
Done. I moved this to a separate method print_results.